### PR TITLE
Include command does not work #177

### PR DESCRIPTION
**Pull request by rachx**
Fixes https://github.com/MarkBind/markbind/issues/177

Not sure if the include/render command was used/useful previously. When it include/render a page, it does not build dynamically included pages (not fixed) or resolve the base url (fixed).

Not sure if we should assume that <file> is within a site folder too. i.e. Might not find a `site.json` from traversing up.

Fixes part of the issue by calculating the root path and user defined variables if the command is used.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -20,6 +23,9 @@ const CLI_VERSION = require('./package.json').version;
 
 const ACCEPTED_COMMANDS = ['version', 'include', 'render', 'init', 'build', 'serve', 'deploy'];
 
+const SITE_CONFIG_NAME = 'site.json';
+const USER_VARIABLES_PATH = '_markbind/variables.md';
+
```
Comment (nicholaschuayunzhi) [inline index.js]: Defining the config name and user variables path here violates DRY principle.

File: index.js [hunk 2]
```diff
@@ -41,7 +47,35 @@ program
...
+    $.root().children().each(function () {
+      const id = $(this).attr('id');
+      const html = $(this).html();
+      userDefinedVariables[id] = html;
+    });
+
+    // This is to prevent the first nunjuck call from converting {{baseUrl}} to an empty string
+    // and let the baseUrl value be injected later.
+    userDefinedVariables.baseUrl = '{{baseUrl}}';
+    config.userDefinedVariablesMap[config.rootPath] = userDefinedVariables;
```
Comment (nicholaschuayunzhi) [inline index.js]: This code seems to be a rewrite of `Site#collectUserDefinedVariablesMap` but for a single site. Would it be possible to extract the parsing of `userDefinedVariables`? Possibly in `util` or in `Site` as a static function? That way we also don't have to import cheerio here as well. Additionally, we can easily write tests for the function.

**Comment by rachx** [comment]
It isn't straightforward to get the constants from the Site file.
I was thinking of extracting all the constants in `index.js`, `site.js` and perhaps even `parser` to a single constant file.

Actually, we were considering deprecating the include/render command since it is almost never used by users and seldom used by developers.

For the second comment, I was thinking about building a site of only one page so the work will be done in Site.js as well.
Also, just to clarify, is it expected that include/render resolves user defined variables/boilerplate files properly but not dynamically included content?

**Comment by acjh** [comment]
> It isn't straightforward to get the constants from the Site file.
I was thinking of extracting all the constants in `index.js`, `site.js` and perhaps even `parser` to a single constant file.

Not all the constants should be extracted. If only used within a file, it's better to be defined there.

> Actually, we were considering deprecating the include/render command since it is almost never used by users and seldom used by developers.

I think it's meant to be useful (especially) for large sites, as an alternative to building an entire site.
But it stops short at serving the page.

> For the second comment, I was thinking about building a site of only one page so the work will be done in Site.js as well.
Also, just to clarify, is it expected that include/render resolves user defined variables/boilerplate files properly but not dynamically included content?

That sounds useful, to provide the complete service. I'm not sure if `include`/`render` should resolve dynamically included content (recursively), as it could lead to a much larger build. Maybe a command line argument?

@Gisonrg Your thoughts?

**Comment by rachx** [comment]
How about additional optional parameters for markbind `build` command instead?
For
* whether to preserve the temp files (file after the include stage)
* the only file path to build (this will be assigned to `this.addressablePages` instead of using `collectAddressablePages()`)
(Potentially we can extend to globs too)

**Comment by acjh** [comment]
> How about additional optional parameters for markbind `build` command instead?
> For
> * whether to preserve the temp files (file after the include stage)

Definitely more accessible to a site author, but I wouldn't expect a site author to use these either.
Let's just fix this [for developers](https://github.com/MarkBind/markbind-cli/wiki/Developer-Guide#markbind-core-library).

> * the only file path to build (this will be assigned to `this.addressablePages` instead of using `collectAddressablePages()`)
(Potentially we can extend to globs too)

We gotta get the title and front matter specified in **site.json** too.
Let's not support this for now.

**Comment by rachx** [comment]
I didn't expect it to be used by site authors. 

It is for developers to use in place of markbind include and render.
With the options, it should be able to achieve the same functions.
The title and front matter should be meaningless for include and render too.

File: lib/markbind/lib/parser.js [hunk 3]
```diff
@@ -50,8 +50,8 @@ function calculateNewBaseUrl(filePath, root, lookUp) {
 }
 
 function calculateBoilerplateFilePath(filePath, config) {
-  const base = calculateNewBaseUrl(filePath, config.rootPath, config.baseUrlMap).relative;
-  return path.resolve(base, BOILERPLATE_DIRECTORY_NAME, path.basename(filePath));
+  const { parent, relative } = calculateNewBaseUrl(filePath, config.rootPath, config.baseUrlMap);
+  return path.resolve(parent, relative, BOILERPLATE_DIRECTORY_NAME, path.basename(filePath));
```
[the remark this answers is not in the document]
Comment (rachx) [reply lib/markbind/lib/parser.js]: Yes (when markbind serve/build was not done from the base directory)
I totally forgot I fixed this here.

[the remark this answers is not in the document]
Comment (acjh) [reply lib/markbind/lib/parser.js]: That was by design, but this seems like an good enhancement.
Can this be a separate PR?

**Comment by acjh** [comment]
@rachx Any update?

**Comment by rachx** [comment]
Force pushed.

**Comment by acjh** [comment]
> Support markbind include/render from the non-root directory in a separate pr?

Yes, why was #202 closed?

**Comment by rachx** [comment]
> Yes, why was #202 closed?

I closed it because it still will not immediately work if markbind include/render was done from a non root directory.
It needs to find the baseUrl first.
I might update it later.

**Comment by acjh** [comment]
@rachx You can merge this :)

**Pull request merged**